Deck list

FabienDeshayes opened the pull request:
This is a draft PR adding the deck list page:
![image](https://user-images.githubusercontent.com/762494/103296972-2dd6df00-49ef-11eb-8f51-24c68ce35348.png)
On mobile:
![image](https://user-images.githubusercontent.com/762494/103296991-3a5b3780-49ef-11eb-9978-095b46b7faa2.png)

This PR includes:
- listing all decks 10 by 10 (load more scrolling to the bottom)
- Filter by name
- Filter by phoenixborn
- List dice and cards (per type) in deck
- Clicking on deck name redirects to `deck/<deck_id>`

Outstanding:
- Cards previews are done using the `card-link` component, which expects the `is_legacy` flag for legacy cards. However cards in legacy decks don't have this flag. Should we add it to the response or change it on the front-end?
- Deck page: a `deck/<deck_id>` endpoint is needed here
- Better responsiveness: should we hide the cards and just keep the number of cards per type when on mobile ? or perhaps having an icon to expand and show deck contents?
- Column for preconstructed decks (do we want it anymore?)
- Check with Reborn decks - this was only tested with legacy ones

And some questions:
- Should we show the number of cards in the deck? Won't it be 30/30 all the time?
- Do we have transparent images of phoenixborn? That would be a good visual addition to the deck list.
- Do we need some color / icons to give some hierarchy to the different card types? At the moment, Ally / Ready Spell / etc. are in a similar font than card themselves.

FabienDeshayes committed:
Setup scaffolding for deck listing

FabienDeshayes committed:
Add search input to deck listing

FabienDeshayes committed:
List decks and link to their details

FabienDeshayes committed:
Add fields to Deck in deck list

FabienDeshayes committed:
Add details of decks in deck list

FabienDeshayes committed:
Style deck list item

FabienDeshayes committed:
Add phoenixborn filter for decks

FabienDeshayes committed:
Add all phoenixborn to phoenixborn picker

FabienDeshayes committed:
Style phoenixborn picker

onecrayon commented on src/components/shared/PhoenixbornPicker.vue:
```diff
@@ -0,0 +1,44 @@
+      @change="$emit('update:filter', $event.target.value)">
+      <option value="" disabled selected>{{ placeholder }}</option>
+      <option value="">All Phoenixborn</option>
+      <option value="james-endersight">Aradel Summergaard</option>
```
Looks like a value copy/paste error. I think this stuff should probably be populated by the server, in any instance, since otherwise it's not possible to add new cards without updating the front-end source code. For instance, when a page with a PB filter is loaded for the first time, it could request a listing of Phoenixborn names/stubs from the Vuex store, and if that data isn't loaded run a quick `/v2/cards?type=phoenixborn` request, then cache the results in the store (that will also make it easy to support legacy Phoenixborn vs. Reborn ones).

FabienDeshayes replied:
This is now done!
Not sure how much it's useful to have it cached though, as anything reload the page entirely. 
But it's a good thing to fetch the list dynamically!

onecrayon commented on src/components/decks/DeckTable.vue:
```diff
@@ -0,0 +1,61 @@
+    Deck,
+  },
+  mounted () {
+    this.debouncedScrollListener = debounce(this.scrollLoadCheck, 100)
```
For the deck listing page, please make it paginated instead of endlessly scrolling (there were a *lot* of decks in the legacy site, so endless scrolling would kind of suck). Optionally, this can just be "next/previous" buttons if calculating out the pages is too complicated with the API return data. I'm not sure how to handle pagination in the query string; maybe don't worry about it for now, since I don't think anyone linked to specific pages ever anyway (given the contents are constantly changing).

FabienDeshayes replied:
I've done a simple Previous / Next implementation for now, but adding the pages shouldn't be too hard. I'll look at it this weekend hopefully.

onecrayon commented on src/components/decks/Deck.vue:
```diff
@@ -0,0 +1,69 @@
+        <span class="text-lg">
+          <card-link :card="deck.phoenixborn"></card-link>
+        </span>
+        <span class="text-sm float-right">
```
If the card count isn't 30, please style the text in an error color or something (this is possible for private deck listings).

FabienDeshayes replied:
Done!

onecrayon commented on src/components/decks/Deck.vue:
```diff
@@ -0,0 +1,69 @@
+      </div>
+      <div class="clear-right">
+        <span class="text-sm">
+          {{ deck.user.username }}
```
Please create a component to output the username that concatenates the username with a lighter gray badge (like "**Skaak**#1234"). Eventually, I'll be swapping this out for a link, so having a component will make life easier for that.

FabienDeshayes replied:
Done, hope it looks fine. If temporary anyway that's probably ok:
![image](https://user-images.githubusercontent.com/762494/103447670-149b9000-4c86-11eb-9b3b-82756a73bd0d.png)

onecrayon reviewed (changes requested):
Apologies for taking a while to get back to you on this stuff! This is generally looking pretty good. Some responses to your questions:

* Let's load maybe 30 decks per page instead of 10 (if that causes a big slow-down, please let me know if it's the front-end or the server holding things up)
* Legacy cards should be marked as such in the server response, so I'll handle that. I think I just forgot to include the property in the schema definition.
* Don't worry about preconstructed decks for now; we can add that later if we need (might just make a filter control for it instead)
* I'll get you Reborn deck data someday... 😔 
* Mentioned this inline, but the number of cards might not be 30 all the time (specifically for viewing private decks)
* I'll be getting you some nice big Phoenixborn images here soon. Just need to process them and get them into the CDN.
* I would love to see the type headings distinguished somehow from the cards themselves. Could tweak the sizing, make one or the other lighter, add the type icons that are already in use across the rest of the site, etc. Feel free to experiment.

Lastly, please add a default style for the HR element in the global styles to make it a nice gray (maybe match the border gray?).

onecrayon commented:
@FabienDeshayes Latest pushes to my `feature/decks-crud` branch should include the `is_legacy` flag for all cards in deck listings. I've got the basic deck creation logic ported, too, so I should hopefully have a new data set for you with Reborn decks soon (that's my goal this weekend, along with getting the Phoenixborn images processed).

FabienDeshayes committed:
Load 30 decks per page

FabienDeshayes committed:
Add style for horizontal separator

FabienDeshayes committed:
Highlight card number if not 30

FabienDeshayes committed:
Add user badge component

FabienDeshayes committed:
Improve card type label with icon

FabienDeshayes committed:
Fix ommission in phoenixborn picker

FabienDeshayes committed:
Add basic pagination

FabienDeshayes committed:
Add page number and page total on deck list page

FabienDeshayes committed:
Load phoenixborn list dynamically

FabienDeshayes commented:
I think the PR is now ready to be reviewed. It probably needs testing with reborn data before merging though.
I was thinking of doing a separate PR for the deck details - is there any example of what that should look like?

onecrayon commented:
@FabienDeshayes I've uploaded the Reborn Phoenixborn images. Small badges for the deck listings are at URLs like:

    https://cdn.ashes.live/images/phoenixborn-badges/aradel-summergaard.jpg

Large size images for the deck detail page are at URLs like:

    https://cdn.ashes.live/images/phoenixborn/aradel-summergaard.jpg

The legacy images are at the same old URLs as they used, but in a legacy folder on the CDN. So for instance:

    https://cdn.ashes.live/legacy/images/cards/aradel-summergaard-large.jpg
    https://cdn.ashes.live/legacy/images/cards/aradel-summergaard-slice.jpg

I will add legacy images for the Phoenixborn that were originally missing due to a lack of high quality images everything after Koji/Astrea), so don't worry about those missing.

And here's an HTML export example of the deck detail page (I should have the endpoint for fetching this data pretty quick, since it shares a lot of logic with the saving endpoint that I'm working on):

[deck-example.zip](https://github.com/onecrayon/ashes.live/files/5761067/deck-example.zip)

I'll pull this down locally to take a look and let you know if I have any other feedback!

FabienDeshayes commented:
It looks great with the images:

Large:
![image](https://user-images.githubusercontent.com/762494/103481212-58e77700-4dd1-11eb-8afb-71b43b481245.png)

Medium:
![image](https://user-images.githubusercontent.com/762494/103481215-6270df00-4dd1-11eb-9f73-981a024f11a9.png)

Mobile:
![image](https://user-images.githubusercontent.com/762494/103481228-71f02800-4dd1-11eb-9483-4655f3f8bff6.png)

I will look at the deck details in a separate PR.

FabienDeshayes committed:
Add phoenixborn badge to deck item in deck list

onecrayon commented:
Nice! If the image is a background image, there's no need to set a minimum height at the large size (it's alright for the image to get cropped). If it's inline, though, don't worry about it. A little extra whitespace isn't going to hurt.

onecrayon commented:
@FabienDeshayes Basic deck fetching logic is now implemented at GET `v2/decks/{deck_id}`. I need to move the deck details into an object property so that I can output the required releases and similar metadata, though. Right now it's kind of a pointless endpoint because you get all the same information as you have already received from the listing.

onecrayon commented:
@FabienDeshayes Added you as a collaborator (I don't know why I forgot to do this initially) so you should be able to create branches on the main repo, assign yourself to issues, etc. I've also re-targeted this PR to a feature branch where we can gather all of your decks work, since I'll need to keep it off the main branch until the API changes are public (anything merged to `main` is immediately deployed to the live site).

onecrayon commented:
Pulled down a local copy of this branch for testing. This is looking great (I particularly like the addition of the icons to the card type headers)! Here's the minor issues and feedback I ran across:

* When updating filtration, the `offset` query parameter needs to be automatically cleared (otherwise the filter results always end up empty, because the offset is typically too high)
* I haven't tracked down what is causing this, but when filtering with a single page of results the page info at the bottom says "Page NaN of 1"
* Please add some more space between the Phoenixborn image and the rest of the content (maybe `mr-4`, or possibly `mr-2` if that's too much at smaller screen sizes)
* The HR element should have the same amount of spacing above and below it (or possibly less below, since the line height on the text under it pushes the content down a bit on its own)
* If possible, I'd love for some spacing above the deck title to cause it to bottom align with the dice (could try using line height for this, though that might make it look terrible when it wraps at smaller sizes)
* For the card listings, I'm curious if `grid-flow-col-dense` would result in good-looking layouts or if it would just ugly things up. When there's six columns at large size things flow kind of weird:
    ![Screen Shot 2021-01-03 at 9 54 11 AM](https://user-images.githubusercontent.com/33501/103485220-a8b24800-4da9-11eb-92e1-8fe28dce5aea.png)
* The addition of the images has broken the smallest size page (dice overflows the width and gets bumped down beneath the image). Probably need to collapse the dice down to a smaller size as you approach the minimum 340px screen, and maybe just automatically drop it under the deck title and left-align pro-actively. The "last updated" and username can also cause weird overflows:
    ![Screen Shot 2021-01-03 at 9 56 21 AM](https://user-images.githubusercontent.com/33501/103485267-fcbd2c80-4da9-11eb-8c1b-a27cef95e19a.png)
* While you're at it, the updated timestamp should probably be gray since it's not very important information.

onecrayon commented:
Lastly, Deep Equal is still a dependency on this branch, but doesn't appear to be in use anywhere anymore.

FabienDeshayes committed:
Tweak deck list following PR comments

FabienDeshayes committed:
Remove deep-equal

FabienDeshayes commented on package-lock.json:
```diff
@@ -1269,6 +1268,12 @@
         "http-errors": "~1.7.2"
       },
       "dependencies": {
+        "deep-equal": {
```
Not sure why this is appearing here as I removed it from package.json and did a clean install. It shows as a transitive dependency as well...

FabienDeshayes commented:
I did some more changes:

* reset offset when Phoenixborn or filter changes. I also fixed an issue where two requests were made sometimes.
* I cannot find how to reproduce the Page NaN of 1. The formula is quite straightforward and to have NaN as a result, offset would have to be NaN as well. 
* the Phoenixborn image is now a background image.
* Spacing on the item should be better
* I couldn't use `grid-flow-col-dense` as cols have different widths and it gets messy. For extra large I've just made it use 6 columns so it all fits in one row - I think it looks better now.
* Overflows should be fixed as I've moved to use flex layouts more. As a result, responsiveness is better too
* Removed deep-equal as well
* Fixed an issue where the Phoenixborn picker would not reset correctly 
* Implemented Clear filters button which was missing

FabienDeshayes committed:
Ensure card link request server data if it only has partial info

FabienDeshayes commented:
@onecrayon do you mind pull the branch and checking the PR again? I'll start a new one for deck details.

onecrayon commented:
@FabienDeshayes Apologies for taking to so long to get to this! Really unproductive week for me, Ashes-wise. This is looking great! I can't reproduce the "NaN" thing anymore, either, so I'm merging this into the base branch to give you something to work off once you get to the next deck stuff.

Thank you!
